### PR TITLE
Add support for OAuth2 authentication and Send on Behalf of

### PR DESCRIPTION
**Pull request by kbussell**
Implements #76

### CODE REVIEW

File: pydocusign/client.py [hunk 1]
```diff
@@ -79,6 +80,14 @@ def __init__(self,
         if not self.app_token:
             self.app_token = os.environ.get('DOCUSIGN_APP_TOKEN', '')
 
+        #: OAuth2 Token.
+        #:
+        #: If not explicitely provided or empty, then ``DOCUSIGN_OAUTH2_TOKEN``
+        #: environment variable, if available, is used.
+        self.oauth2_token = oauth2_token
+        if not self.oauth2_token:
+            self.oauth2_token = os.environ.get('DOCUSIGN_OAUTH2_TOKEN', '')
```
Comment (jpic) [inline pydocusign/client.py]: In which case would this be an env var ? For debugging purposes ?

Comment (kbussell) [inline pydocusign/client.py]: Yeah, this was done to be analogous to the other authentication settings.

**Comment by zebuline** [comment]
Thx @kbussell  for this PR.
Could you please rebase it ?

**Comment by kbussell** [comment]
Done. Sorry for the delay.

**Comment by zebuline** [comment]
@kbussell the build "push" failed (don't look at the job "pr", it failed because encrypted vars are missing - it is a feature from travis, cf https://github.com/novafloss/django-docusign/pull/54#issuecomment-235254013)

```
======================================================================

1) ERROR: DocuSignClient.login_information() populates account information.

----------------------------------------------------------------------

   Traceback (most recent call last):

    tests.py line 111 in test_login_information

      result = docusign.login_information()

    pydocusign/client.py line 239 in login_information

      data = self.get(url, headers=headers)

    pydocusign/client.py line 220 in get

      return self._request(method='GET', *args, **kwargs)

    pydocusign/client.py line 212 in _request

      raise exceptions.DocuSignException(msg)

   DocuSignException: DocuSign request failed: GET https://demo.docusign.net/restapi/v2/login_information returned code 401 while expecting code 200; Message: {

     "errorCode": "AUTHORIZATION_INVALID_TOKEN",

     "message": "The access token provided is expired, revoked or malformed."

   } ; 

   -------------------- >> begin captured logging << --------------------

   requests.packages.urllib3.connectionpool: INFO: Starting new HTTPS connection (1): demo.docusign.net

   requests.packages.urllib3.connectionpool: DEBUG: "GET /restapi/v2/login_information HTTP/1.1" 401 127

   pydocusign.client: ERROR: DocuSign request failed: GET https://demo.docusign.net/restapi/v2/login_information returned code 401 while expecting code 200; Message: {

     "errorCode": "AUTHORIZATION_INVALID_TOKEN",

     "message": "The access token provided is expired, revoked or malformed."

   } ; 

   --------------------- >> end captured logging << ---------------------

======================================================================

2) ERROR: test_sobo_with_oauth2 (tests.SOBOTestCase)

----------------------------------------------------------------------

   Traceback (most recent call last):

    tests.py line 329 in test_sobo_with_oauth2

      headers = client._create_envelope_from_document_request(envelope)['headers']

   KeyError: 'headers'

======================================================================

3) ERROR: test_sobo_with_regular_auth (tests.SOBOTestCase)

----------------------------------------------------------------------

   Traceback (most recent call last):

    tests.py line 351 in test_sobo_with_regular_auth

      headers = client._create_envelope_from_document_request(envelope)['headers']

   KeyError: 'headers'
```

**Comment by zebuline** [comment]
@kbussell still 2 errors: https://travis-ci.org/novafloss/pydocusign/jobs/147766941

**Comment by kbussell** [comment]
Tests finally passed, so squashed commits

**Comment by zebuline** [comment]
Yeah, well done !
Thanks @kbussell

**Pull request merged**